Extra hashtags: automatic and suggested hashtags for proposals

leio10 opened the pull request:
#### :tophat: What? Why?

We need to define several participatory processes where we want to allow users to:
* assign proposals to several categories
* search proposals by category from different participatory processes at once

We can't use Decidim's categories, since proposals can't be assigned to several categories and can't be shared between different participatory process. So, we are going to use hashtags for that purpose. But to do this, we need to allow admins to guide users on which are the defined "categories".

First, this PR defines an additional use of hashtags: extra hashtags. These are regular hashtags added to a content, but with an special character to mark them as "extra" for that particular occasion. While `gid://decidim-development-app/Decidim::Hashtag/1/Hashtag` adds a reference to a hashtag,  `gid://decidim-development-app/Decidim::Hashtag/1/_Hashtag` adds a reference to the same hashtag, but as an "extra" hashtag. These kind of references can be easily hidden when rendered and also can be retrieved separately as an array.

Then, this PR implements two features using extra hashtags, both in the Proposals component: automatic and suggested hashtags. Automatic hashtags are hashtags that are assigned to all proposals created in a proposals component. This can be useful to add additional information to proposals to improve global search results.
Suggested hashtags allow to define a set of predefined hashtags that can be offered to users to improve organization of the proposals, also to improve global search results.

#### :pushpin: Related Issues

#### :clipboard: Subtasks
- [x] Add `CHANGELOG` entry
- [ ] Add documentation regarding the feature 
- [x] Add tests

### :camera: Screenshots (optional)
![imagen](https://user-images.githubusercontent.com/453545/49100608-971a2000-f274-11e8-8401-97d5f8b9e580.png)
![imagen](https://user-images.githubusercontent.com/453545/49100627-a4370f00-f274-11e8-922e-ef853c01f127.png)
![imagen](https://user-images.githubusercontent.com/453545/49100689-bf098380-f274-11e8-878f-18cd2415cbba.png)

![peek 27-11-2018 18-13](https://user-images.githubusercontent.com/453545/49100023-38a07200-f273-11e8-9a0a-1defdd8ea542.gif)

leio10 commented:
@decidim/product I'm still have work in this PR (tests, documentation, changelog), but I'd like to know if you think that this feature could be helpful for Decidim, or if you have some feedback about how is implemented.

mrcasals commented:
Wow @leio10, it looks awesome to me! 👏 👏 👏 👏 

Let's wait for @decidim/product's answer! 😄

arnaumonty commented:
@leio10 Looks great! It solve the problems that we have when we want to assign more than one category/tag. I see important to be enabled/disabled from admin proposals component. But it looks so useful.

Cc @decidim/product

carolromero commented:
@leio10 looks awesome!!

arnaumonty commented:
> Regarding @arnaumonty comment, just leaving empty any of the fields will disable that feature. Do you think that will be clear enough for admin users?

@leio10 I think that is clear. We should document in our admin docs, but I see it enough.

Just a question. How it works the suggested ones?

leio10 commented:
@arnaumonty You can see it working [here](https://user-images.githubusercontent.com/453545/49100023-38a07200-f273-11e8-9a0a-1defdd8ea542.gif):
* an admin defines a list of hashtags separated by spaces
* when a user completes a proposal, a checkbox for each suggested hashtag appears
* when saving the proposal, selected hashtags are added to the proposal body as other hashtags (after a line break)
* when editing the proposal, suggested hashtags are removed from the body, and their checkboxes are checked

Something very similar occurs with the automatic hashtags, with the difference that users can't uncheck those checkboxes.

leio10 commented:
@arnaumonty @carolromero I guess that collaborative drafts should also use these features if are enabled for proposals, do you agree?

arnaumonty commented:
@leio10 It sounds good.

leio10 commented:
Ok, I've added hashtags to collaborative drafts (there were a lot of work pending there :sweat_smile:) and then added this feature to collaborative drafts. I'll add some tests and docs and this will be ready to be reviewed.

mrcasals commented:
>  I've added hashtags to collaborative drafts

@decidim/lot-mods pinging you so you're aware of this, just in case this affects your work 😄

mrcasals commented:
@leio10 #4608 should fix the error you're encountering on `decidim-core`!

leio10 commented:
Ok, @mrcasals I think this is ready to review!! :smile:

Sorry about the size of this PR, this is a brief summary of my journey:
* I added extra hashtags to proposal creation wizard
* I tried to add extra hashtags to collaborative drafts, but collaborative drafts seems like an outdated copy of proposals, without support for hashtags or the very nice forms refactoring applied on #3744.
* I added hashtags support to collaborative drafts
* I refactored collaborative drafts forms
* I added extra hashtags to collaborative drafts
* I added some tests for the added features
* I added extra hashtags to admin proposals creation, because I realized that admin and front proposals forms share the same tests
* I added collaborative drafts creation system tests, because I realized that collaborative drafts didn't have creation system tests similar to proposals test

Finally, just a thought: having collaborative drafts as a copy of proposals leads to a lot of code duplication. IMHO we should think a way to reuse code from proposals that would help to reduce mainteinance effort.

oriolgual commented:
> Finally, just a thought: having collaborative drafts as a copy of proposals leads to a lot of code duplication. IMHO we should think a way to reuse code from proposals that would help to reduce mainteinance effort.

I agree, it should have been done using modules to share the business logic but for time constraints it wasn't done like this.

leio10 commented:
@mrcasals @oriolgual hi, who can review this? :grimacing: 
I can rebase it again, but it is so big that every commit to master add conflicts, so I prefer to wait that it is reviewed.
Regarding documentation, I don't know if there is any place where I can add an explanation for this.

oriolgual commented on decidim-proposals/app/forms/decidim/proposals/admin/proposal_form.rb:
```diff
@@ -92,6 +117,12 @@ def scope_belongs_to_participatory_space_scope
+        def ordered_hashtag_list(string)
+          return [] unless string
+
+          string.split.reject(&:blank?).uniq.sort_by(&:parameterize)
```
```suggestion
          string.to_s.split.reject(&:blank?).uniq.sort_by(&:parameterize)
```

This way you don't need the early return

oriolgual commented on decidim-proposals/app/forms/decidim/proposals/proposal_form.rb:
```diff
@@ -25,6 +26,12 @@ class ProposalForm < Decidim::Proposals::ProposalWizardCreateStepForm
+      def map_model(model)
+        super
+
+        @hashtags_suggested = Decidim::ContentRenderers::HashtagRenderer.new(model.body).extra_hashtags.map(&:name).map(&:downcase)
```
Same changes to the other form

oriolgual commented on decidim-proposals/app/models/decidim/proposals/collaborative_draft.rb:
```diff
@@ -27,6 +27,8 @@ class CollaborativeDraft < Proposals::ApplicationRecord
                class_name: "Decidim::User",
                foreign_key: :decidim_user_id
 
+      geocoded_by :address, http_headers: ->(collaborative_draft) { { "Referer" => collaborative_draft.component.organization.host } }
```
Why this?

leio10 replied:
When refactored collaborative drafts, I've found that geocoding feature was failing for collaborative drafts, trying to call a `geocoded` method. I've compared with proposals and found that this line was applied there but not here. After adding it, it worked well. Maybe I should add a regression test... :grimacing:

leio10 replied:
Ok, I added the regression test [here](https://github.com/decidim/decidim/pull/4585/commits/0481b100f635ee09d7e586f8cca95320c6f45c77) (and checked that it fails without the `geocoded_by` line) :v:

oriolgual commented on decidim-proposals/app/presenters/decidim/proposals/proposal_presenter.rb:
```diff
@@ -34,24 +34,14 @@ def display_mention
       end
 
-      def title
+      def title(links: false, extras: true)
```
Param docs?

oriolgual replied:
Also, about the defaults, is it more common or expected to have these as true or false?

leio10 replied:
Yes, it is much common to use params as given (don't render hashtag links, but include extras). That's why I leave it that way

oriolgual commented on decidim-proposals/app/presenters/decidim/proposals/proposal_presenter.rb:
```diff
@@ -34,24 +34,14 @@ def display_mention
-      end
-
-      def html_body
+      def body(links: false, extras: true)
```
Param docs?

oriolgual commented on decidim-proposals/config/locales/ca.yml:
```diff
@@ -421,6 +419,7 @@ ca:
           error: Hi ha hagut un problema en crear aquests esborranys col·laboratius
           success: S'ha creat un esborrany col·laboratiu.
         edit:
+          attachment_legend: "(Opcional) Afegeix un fitxer adjunt"
```
We shouldn't change non-english locales, otherwise we'll have merge conflicts with PRs coming from Crowdin.

oriolgual commented on decidim-proposals/config/locales/en.yml:
```diff
@@ -15,6 +15,8 @@ en:
         category_id: Category
         has_address: Has address
+        hashtags_auto: Hashtags automatically added
+        hashtags_suggested: Hashtags suggested
```
```suggestion
        suggested_hashtags: Suggested hashtags
```

oriolgual commented on decidim-proposals/config/locales/en.yml:
```diff
@@ -97,6 +99,8 @@ en:
             creation_enabled: Proposal creation enabled
             endorsements_blocked: Endorsements blocked
             endorsements_enabled: Endorsements enabled
+            hashtags_auto: Hashtags added to all proposals
```
After reading the description, maybe `default_hashtags` would be a better naming?

leio10 replied:
Mmm, I think that a default hashtag is something that is included at first, but that you can modify. In this case, these hashtags are fixed, they can't be removed from the proposal. This doesn't add too much value for the proposal inside the process (as all proposals have the same hashtags), but its a nice way to group proposals created in different participatory processes. For example, we are planning to use this to group proposals created in participatory processes at town level or at province level.
I can think two possible names for this:
* automatic_hashtags
* fixed_hashtags

I'd go with the first, but maybe you can think a better name

oriolgual replied:
`automatic_hashtags` sounds good!

oriolgual commented on decidim-proposals/config/locales/en.yml:
```diff
@@ -410,8 +414,6 @@ en:
           title: Similar collaborative drafts
         complete:
-          attachment_legend: "(Optional) Add an attachment"
-          select_a_category: Please select a category
```
Why this?

leio10 replied:
It is explained in [this commit](https://github.com/decidim/decidim/pull/4585/commits/1216a62d3a3d1185cee617e9a37637963122b8ae): I've refactored collaborative drafts templates using a partial, so I reduced a bit the need of translating the same text several times. Also, I changed some translation keys paths included in that partial, that is way I touched all the locales files, to reuse existing translation instead of translating it again.

leio10 commented:
@oriolgual Thanks for spending your time to review this :heart:

I think that every issues are fixed, except maybe the `hashtags_auto` naming. Please, let me know if there is anything else that must be addressed before merging :smiley_cat:

oriolgual commented on decidim-proposals/app/commands/decidim/proposals/hashtags_methods.rb:
```diff
@@ -0,0 +1,36 @@
+
+module Decidim
+  module Proposals
+    # A module with all the attachment common methods for proposals
```
```suggestion
    # A module with all the hash tags common methods for proposals
```

mrcasals commented:
Wohooooo! Thanks for your contributions @leio10! ❤️

leio10 commented:
:tada: :tada: :tada: Yeah!! Thanks for helping me with this!! :heart_eyes:

tramuntanal commented:
Hi @leio10 do we have any chance to have some documentation on this. It seems a great feature :heart:  but it would be great to know how it works :woman_shrugging:

leio10 commented:
Hi @tramuntanal, I agree with you, but I didn't found a place to put that.

Now I'm thinking that we could create a hashtags.md file at [/docs/advanced](https://github.com/decidim/decidim/tree/master/docs/advanced) and explain the hashtags feature there. Something like this:

Hashtags are useful to relate resources between different components and participatory processes.

When developing a component, you should consider this to add hashtags support.
* There is a `HashtagParser` class in `Decidim::Core` that should be used to process content that can have hashtags:
```
(example of use)
```
* When parsing the content, the hashtag `#MeToo` is converted to `gid://decidim-development-app/Decidim::Hashtag/1/MeToo` and should be stored in that format in the database
* `HashtagParser` can be configured to generate extra hashtags, that add that information to the formatted content (`gid://decidim-development-app/Decidim::Hashtag/1/_MeToo`)
* Hashtags are stored all in lowercase in the database. Case info is stored in the content in each hashtag reference.
* There is a `HashtagRenderer` class in `Decidim::Core` that should be used to show content with hashtags:
```
example of use
```
* `HashtagRenderer` can be configured to render links to hashtag search or to hide extra hashtags.
* Extra hashtags can be retrieved separately from the renderer.

Currently, Meetings and Proposals modules support hashtags. In the case of the Proposals component, it allows to add hashtags to proposals and collaborative drafts title and body. Also, it uses extra hashtags to allow admin users to configure predefined hashtags that will be added to the proposals automatically or by user interaction.

What do you think? Is this what you have in mind?
